src/Data/Array/Mutable/Linear.hs

```diff
@@ -69,17 +75,52 @@ read = Linear.toLinear readUnsafe
       | indexInRange size ix = (arr, readMutArr mutArr ix)
       | otherwise = error "Read index out of bounds."
 
+-- | Using first element as seed, resize to a constant array
 resize :: HasCallStack => Int -> Array a #-> Array a
 resize newSize (Array _ mutArr) =
-  Array newSize (resizeMutArr mutArr newSize)
+  Array newSize (newMutArr newSize (readMutArr mutArr 0))
 
--- Resize with a seed value for the newly allocated elements
+-- | Resize to a new constant array given a seed value
 resizeSeed :: HasCallStack => Int -> a -> Array a #-> Array a
-resizeSeed newSize seed (Array _ mutArr) =
-  Array newSize (resizeMutArrSeed mutArr seed newSize)
+resizeSeed newSize seed (Array _ _) =
+  Array newSize (newMutArr newSize seed)
+
+-- XXX: Replace with toVec
+toList :: Array a #-> (Array a, [a])
+toList = Linear.toLinear unsafetoList
+  where
+    unsafetoList :: Array a -> (Array a, [a])
+    unsafetoList arr@(Array size _) = toListWalk 0 size arr []
+      where
+        toListWalk :: Int -> Int -> Array a -> [a] -> (Array a, [a])
+        toListWalk ix size arr accum
+          | ix == size = (arr, accum)
+          | otherwise =
+            toListWalk (ix + 1) size arr ((snd $ read arr ix) : accum)
+
+-- # Instances
+-------------------------------------------------------------------------------
+
+-- # This instance is only provided for debugging. It's not pretty.
+instance Show a => Show (Array a) where
+  show (Array size mutArray) = showArr 0 size mutArray
+    where
+      showArr n size mutArray
+        | n == 0 =
+          "\n[\n" ++ show (readMutArr mutArray 0) ++ "\n"
+            ++ showArr (n + 1) size mutArray
+        | (n + 1) == size =
+          show (readMutArr mutArray n) ++ "\n]\n"
+        | otherwise =
+          show (readMutArr mutArray n) ++ "\n"
+            ++ showArr (n + 1) size mutArray
```

Comment:
```suggestion
  show (Array size mutArray) = "\n[\n" ++ showArr 0 size mutArray ++ "]\n"
    where
      showArr n size mutArray
        | n == size = ""
        | otherwise = show (readMutArr mutArray n) ++ "\n" ++ showArr (n + 1) size mutArray
```